src/cryptography/hazmat/backends/openssl/x509.py

```diff
@@ -91,3 +91,45 @@ def _parse_asn1_time(self, asn1_time):
             )
         ).decode("ascii")
         return datetime.datetime.strptime(time, "%Y%m%d%H%M%SZ")
+
+    @property
+    def issuer(self):
+        issuer = self._backend._lib.X509_get_issuer_name(self._x509)
+        assert issuer != self._backend._ffi.NULL
+        return self._build_x509_name(issuer)
+
+    @property
+    def subject(self):
+        subject = self._backend._lib.X509_get_subject_name(self._x509)
+        assert subject != self._backend._ffi.NULL
+        return self._build_x509_name(subject)
+
+    def _build_x509_name(self, x509_name):
+        count = self._backend._lib.X509_NAME_entry_count(x509_name)
+        attributes = []
+        for x in range(count):
+            entry = self._backend._lib.X509_NAME_get_entry(x509_name, x)
+            obj = self._backend._lib.X509_NAME_ENTRY_get_object(entry)
+            assert obj != self._backend._ffi.NULL
+            data = self._backend._lib.X509_NAME_ENTRY_get_data(entry)
+            assert data != self._backend._ffi.NULL
+            buf = self._backend._ffi.new("unsigned char **")
+            res = self._backend._lib.ASN1_STRING_to_UTF8(buf, data)
+            assert buf[0] != self._backend._ffi.NULL
+            buf = self._backend._ffi.gc(
+                buf, lambda buf: self._backend._lib.OPENSSL_free(buf[0])
+            )
+            value = self._backend._ffi.buffer(buf[0], res)[:].decode('utf8')
+            buf_len = 50
```

Comment:
Perhaps `X509_NAME_MAXLEN` is what we should be using?

On Fri, Feb 13, 2015 at 6:47 PM, Paul Kehrer <notifications@github.com>
wrote:

> In src/cryptography/hazmat/backends/openssl/x509.py
> <https://github.com/pyca/cryptography/pull/1612#discussion_r24709108>:
>
> > +        count = self._backend._lib.X509_NAME_entry_count(x509_name)
> > +        attributes = []
> > +        for x in range(count):
> > +            entry = self._backend._lib.X509_NAME_get_entry(x509_name, x)
> > +            obj = self._backend._lib.X509_NAME_ENTRY_get_object(entry)
> > +            assert obj != self._backend._ffi.NULL
> > +            data = self._backend._lib.X509_NAME_ENTRY_get_data(entry)
> > +            assert data != self._backend._ffi.NULL
> > +            buf = self._backend._ffi.new("unsigned char **")
> > +            res = self._backend._lib.ASN1_STRING_to_UTF8(buf, data)
> > +            assert buf[0] != self._backend._ffi.NULL
> > +            buf = self._backend._ffi.gc(
> > +                buf, lambda buf: self._backend._lib.OPENSSL_free(buf[0])
> > +            )
> > +            value = self._backend._ffi.buffer(buf[0], res)[:].decode('utf8')
> > +            buf_len = 50
>
> I'm mistaken. Since this is actually an ASN1_OBJECT (not a string) we
> can't use that function. OBJ_obj2txt converts it to the dotted string
> form, but there doesn't appear to be a way to obtain the number of bytes
> this will actually consume.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pyca/cryptography/pull/1612/files#r24709108>.
>



-- 
"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: 125F 5C67 DFE9 4084